clippy_lints/src/loops/manual_find.rs

```diff
@@ -0,0 +1,166 @@
+use rustc_errors::Applicability;
+use rustc_hir::{
+    lang_items::LangItem::{OptionNone, OptionSome},
+    BindingAnnotation, Expr, ExprKind, Node, Pat, PatKind, QPath, Stmt, StmtKind,
+};
+use rustc_lint::LateContext;
+use rustc_span::source_map::Span;
+use rustc_span::symbol::Ident;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    pat: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    body: &'tcx Expr<'_>,
+    span: Span,
+    expr: &'tcx Expr<'_>,
+) {
+    let inner_expr = peel_blocks_with_stmt(body);
+    // Check for the specific case that the result is returned and optimize suggestion for that (more
+    // cases can be added later)
+    if_chain! {
+        if let Some(higher::If { cond, then, r#else: None, }) = higher::If::hir(inner_expr);
+        if let Some(ident) = get_ident(pat);
+        if let ExprKind::Block(block, _) = then.kind;
+        if let [stmt] = block.stmts;
+        if let StmtKind::Semi(semi) = stmt.kind;
+        if let ExprKind::Ret(Some(ret_value)) = semi.kind;
+        if let ExprKind::Call(Expr { kind: ExprKind::Path(ctor), .. }, [inner_ret]) = ret_value.kind;
+        if is_lang_ctor(cx, ctor, OptionSome);
+        if let Some(i) = ident_of(inner_ret);
+        if i == ident;
+        if let Some((first_ret, last_ret)) = extract_rets(cx, expr);
+        // Some(binding)
+        if let ExprKind::Call(_, [binding]) = first_ret.kind;
+        then {
+            let mut applicability = Applicability::MachineApplicable;
+            let mut snippet = make_iterator_snippet(cx, arg, &mut applicability);
+            let is_ref_to_binding =
+                matches!(pat.kind, PatKind::Ref(inner, _) if matches!(inner.kind, PatKind::Binding(..)));
+            if !(matches!(pat.kind, PatKind::Binding(..)) || is_ref_to_binding) {
+                snippet.push_str(
+                    &format!(
+                        ".map(|{}| {})",
+                        snippet_with_applicability(cx, pat.span, "..", &mut applicability),
+                        snippet_with_applicability(cx, binding.span, "..", &mut applicability),
+                    )[..],
+                );
+            }
+            let mut borrows = 0;
+            match cond.kind {
+                ExprKind::Binary(_, l, r) => {
+                    for side in [l, r] {
+                        if !matches!(
+                            side.kind,
+                            ExprKind::MethodCall(_, [caller, ..], _) if ident_in_expr(cx, ident, caller)
+                        ) && ident_in_expr(cx, ident, side)
+                        {
+                            borrows = 1;
+                        }
+                    }
+                },
+                ExprKind::Call(..) => borrows = 1,
+                ExprKind::MethodCall(_, [caller, ..], _) => {
+                    if !ident_in_expr(cx, ident, caller) {
+                        borrows = 1;
+                    }
+                },
+                _ => {},
+            }
```

Comment:
I have found quite the spanner to throw in the works:

```rust
fn vec_string(strings: Vec<String>) -> Option<String> {
    for s in strings {
        if s == String::new() {
            return Some(s);
        }
    }
    None
}
```

both `|s|` and `|&s|` would result in errors:

```
error[E0277]: can't compare `&std::string::String` with `std::string::String`
  --> t.rs:25:36
   |
25 |     strings.into_iter().find(|s| s == String::new())
   |                                    ^^ no implementation for `&std::string::String == std::string::String`
   |
   = help: the trait `std::cmp::PartialEq<std::string::String>` is not implemented for `&std::string::String`
```
```
error[E0507]: cannot move out of a shared reference
  --> t.rs:25:31
   |
25 |     strings.into_iter().find(|&s| s == String::new())
   |                               ^-
   |                               ||
   |                               |data moved here
   |                               |move occurs because `s` has type `std::string::String`, which does not implement the `Copy` trait
   |                               help: consider removing the `&`: `s`
```

I believe the only way to create a correct suggestion here would be

```rust
strings.into_iter().find(|s| *s == String::new())
```

Reply:
One thing that you could do is check if `inner_ret` is `Copy`, in that case you can add the extra borrow unconditionally and the body of `find` can be unchanged

For the non `Copy` case it's more tricky, I *think* turning every use of `ident` into `(*ident)` would be correct, though it would be a fair bit of work creating such a snippet. I'd need to think a bit longer on the cases where you can leave it as `ident`

Another option would be to have an `Applicability::HasPlaceholders` suggestion for the non-`Copy` case, e.g. for the strings case suggesting

```rust
strings.into_iter().find(|s| { ... })
```

Reply:
I went with the `Applicability::HasPlaceholders` implementation for now, but I can change it to dereference each `ident`. The only way I can think to do that would be to walk the tree and add partial snippets with asterisks and parentheses in between.

As for the cases that don't have to be dereferenced, all I can think of so far is a method called on `ident`, but I'm probably missing some.